### DISCUSSION TITLE
Seeking possible causes for an assertion error in socksm.c at line 600: hdr.pkt_type == MPIDI_NEM_TCP_SOCKSM_PKT_ID_INFO ||  ...

### DISCUSSION THREAD

**Discussion by tarunkumar987** [pmodels/mpich#7052] [General]
We are hitting following assertion:
Assertion failed in file ...nemesis/netmod/tcp/socksm.c at line 600: hdr.pkt_type == MPIDI_NEM_TCP_SOCKSM_PKT_ID_INFO ||  ...
I looked at the assert and it looks like this in file socksm.c
   MPIU_Assert(hdr.pkt_type == MPIDI_NEM_TCP_SOCKSM_PKT_ID_INFO ||
		hdr.pkt_type == MPIDI_NEM_TCP_SOCKSM_PKT_TMPVC_INFO);

We have tried multiple cores from 2 to 8 and the behaviour is same. Also process is aborted and a message appear to suggest that. Mostly it's process 0 which is aborted but I have seen other processes as well reporting the crash. We are using mpich-3.2.1. I'm trying to understand possible causes for this error? I have explored the forum and no possible causes, like machine going out of memory etc are applicable here. Please suggest.

**Comment by hzhou** [comment]
Does it fail for all applications? If not, can you find a small reproducer?

**Reply by tarunkumar987** [reply]
No it doesn't fail for all the applications. It's the first time we have seen this error in the tool. It currently fails in a customer environment for a big design so there is no easy way to have a small test case. We are currently trying to isolate the issue. Any pointers to what can be tried to isolate?

**Comment by hzhou** [comment]
Is there other activity going on during the run, for example, a port scan or some other application that is trying to make connection on a seemingly random port? The assertion is part of a initial handshake check and it appears it is receiving an unrecognized packet after the connection.

**Reply by tarunkumar987** [reply]
I'm checking that with customer. Meanwhile, can you suggest an experiment doing that? I can try that at my end in a small test case.

**Reply by tarunkumar987** [reply]
https://github.com/pmodels/mpich/issues/2106

Is it something you are suggesting?

**Reply by tarunkumar987** [reply]
I tried above experiemtn but it reported assert at a different place. We are using mpiexec and I could not find a way to pass port range as specified in mpirun. I assumed that's an env variable so i set it as an environment varianble. I ran mpiexec and nmap on separate shells and it gave me following assertion(along with few other errors):

HYD_pmcd_pmiserv_proxy_init_cb (/tmp/mpich//mpich-3.2.1/src/pm/hydra/pm/pmiserv/pmiserv_cb.c:452): assert (!closed) failed

**Reply by hzhou** [reply]
The port scan will fail at hydra first before reaching the netmod. We have strengthened the handshake protocol in hydra since then. I don't think you have a portscan issue, but I think you have a similar cross-connection issue. Hope that helps in your effort of isolating the issue. 

Is upgrading MPICH not an option for you?

**Reply by tarunkumar987** [reply]
Thanks for your input. Unfortunately, upgrading MPICH is not an option at this point.

**Reply by tarunkumar987** [reply]
I stil have a confusion about why this particular assertion is even hitting that too when the application is already running for several hours.  This application is only utilizing multiple cores and not machines. So I'm not sure why this particular handshake happened for single machine run. Can you please reflect?

**Reply by hzhou** [reply]
This is a single-machine run? ;Then indeed, I wonder why it hit the TCP paths at all

**Reply by tarunkumar987** [reply]
Yes. In fact, the tool is not configured to run on multi machine mode. It has to be a single machine run utilizing multiple cores. Is there a configuration setting we might be missing?

**Reply by tarunkumar987** [reply]
We got the confirmation from the customer that there was a security applicaton which was somehow killing the application as the application(and it's executables) were not whitelisted.

I have couple of questions:

1) Is the latest stable release of mpi4.2.2 stable against such instances? Is single machine run still require TCP ports or is it done using shared memory?
2) Is there anything that we can do while configuring mpich 3.2.1(the current version we are using) to make the application more robust against such instances? As I said it's a single machine run, so I'm still not sure why TCP ports are involved.

Thanks
Tarun

**Reply by hzhou** [reply]
Yes, try mpich-4.2.2. By default, mpich v4 series use completely different design (`ch4` vs `ch3`), so it may work better in your case. I need know how exactly the security application is killing the application to comment on whether ch4 would be immune. But, give it a try.

> Is there anything that we can do while configuring mpich 3.2.1(the current version we are using) to make the application more robust against such instances? As I said it's a single machine run, so I'm still not sure why TCP ports are involved.

Try set `MPIR_CVAR_NEMESIS_TCP_NETWORK_IFACE=lo` see if it helps.

**Reply by tarunkumar987** [reply]
Thanks for your reply.

Regarding setting MPIR_CVAR_NEMESIS_TCP_NETWORK_IFACE=lo, is this env variable required while configuring/building mpich? or is it required to be specifed after mpicexec has been built? When we tried it during runtime is was conficting with another env variable MPIR_CVAR_CH3_INTERFACE_HOSTNAME that we did not explicity set(maybe it got set while configuring).

Further, we got information about security utility interferring with mpich. It is a scanner daemon that tries various SSL ciphers and invalid packets on every open port it finds on a host to try to hack in to any insecure webserver.  Whether it feels like a webserver or not. Overall our customer expectation is that application(mpich) captures any bad input(packets) and handle that gracefully by SSL handling code or elsewhere.  Also it was suggested to keep up to date with latest OpenSSL libraries if used by mpich.

My question is whether current stable release of mpich(4.2.2?) handles such bad inputs and is immune to such scans? 

I have another question on using MPICH_PORT_RANGE. It was also considered whether it can be known in advance the port number acquired by mpich and the executaable it spawns. We found this env variable in the docs but not sure if it works reliably. Do you have any comment of it's usage?

Let me know if you have any further questions.

**Reply by hzhou** [reply]
> Regarding setting MPIR_CVAR_NEMESIS_TCP_NETWORK_IFACE=lo, is this env variable required while configuring/building mpich? or is it required to be specifed after mpicexec has been built? When we tried it during runtime is was conficting with another env variable MPIR_CVAR_CH3_INTERFACE_HOSTNAME that we did not explicity set(maybe it got set while configuring).

I see. Try `mpiexec -hosts localhost` option when launching jobs. That should set `MPIR_CVAR_CH3_INTERFACE_HOSTNAME ` to `localhost` which is equivalent to forcing the loopback interface. My hypothesis is the loopback will be isolated from the security app scan.

> Further, we got information about security utility interferring with mpich. It is a scanner daemon that tries various SSL ciphers and invalid packets on every open port it finds on a host to try to hack in to any insecure webserver. Whether it feels like a webserver or not. Overall our customer expectation is that application(mpich) captures any bad input(packets) and handle that gracefully by SSL handling code or elsewhere. Also it was suggested to keep up to date with latest OpenSSL libraries if used by mpich.

I see. This is the exact same issue as the port scan from nmap then. 

MPICH does not use openssl.

> My question is whether current stable release of mpich(4.2.2?) handles such bad inputs and is immune to such scans?

I believe the latest release does not have this issue.

> I have another question on using MPICH_PORT_RANGE. It was also considered whether it can be known in advance the port number acquired by mpich and the executaable it spawns. We found this env variable in the docs but not sure if it works reliably. Do you have any comment of it's usage?

You can set this environment variable so mpich will only use ports in this range. If you can configure your security application to not to scan the same range, that is also a potential solution.

**Reply by tarunkumar987** [reply]
Thanks for your reply. "mpiexec -hosts localhost ..." did not really work. I checked the tcp ports using netstat and I still see tcp ports for each of the processes spawned.

**Reply by tarunkumar987** [reply]
Further, does nmap also send bad packets? AFAIK it's a port scanner.  We reproduced locally another crash with nmap but that did not appear at customer end. Following is the crash that we saw with nmap locally:
HYD_pmcd_pmiserv_proxy_init_cb (/tmp/mpich//mpich-3.2.1/src/pm/hydra/pm/pmiserv/pmiserv_cb.c:452): assert (!closed) failed

Customer only observed this crash in netmod as it found illegal header type in the packet.

**Reply by tarunkumar987** [reply]
We could reproduce this issue locally. Basically we observed two crashes depending on which tcp port is scanned by nmap. In general, we have tcp ports created for mpiexec and our own execs. If nmap is run on tcp port for mpiexec then we see this crash:
HYD_pmcd_pmiserv_proxy_init_cb (/tmp/mpich//mpich-3.2.1/src/pm/hydra/pm/pmiserv/pmiserv_cb.c:452): assert (!closed) failed

However if nmap scans tcp port for our execuaable then we see the orginal assert as follows:
MPIU_Assert(hdr.pkt_type == MPIDI_NEM_TCP_SOCKSM_PKT_ID_INFO ||
hdr.pkt_type == MPIDI_NEM_TCP_SOCKSM_PKT_TMPVC_INFO)

We are looking for both short term and long term solutions. Short term solution since customer is not willing to whitelist our tool from the security app running nmap for long. Long term solution is to move to later(latest) version of mpich as it will go through configuration and testing cycle. For short term, we are exploring certain hotfixes to avoid this assert.

I see following discussion in this thread(https://github.com/pmodels/mpich/issues/6010):

![image](https://github.com/user-attachments/assets/2841a7ba-dfb8-4645-b417-59ed4da0e0ab)

Is this hotfix safe enough? is this fix present in any later releases? Anything else we can attempt?

**Reply by hzhou** [reply]
The hotfix effectively ignores unrecognized connection and simply closes the connection. It should work if you assume all such unrecognized packets comes from benign port scan thus safe to ignore.

**Reply by tarunkumar987** [reply]
For us, as long as packets sent between partitions(or all processes including mpiexec and our executables are recognized) it really won't matter. Can we assume that all other sort of packets are unrecognized/unwanted and are safe to ignore?

**Reply by hzhou** [reply]
Since you are running on a local node, there isn't any communication using tcp/ip, so it should be safe to ignore all such packets.
